skyjo_immutable.py

```diff
@@ -70,6 +70,55 @@ def visible_indices(self) -> list[tuple[int, int]]:
             for i in np.where((self.is_cleared == 0) & (self.is_flipped == 1))[0]
         ]
 
+    @property
+    def valid_discard_and_flip_actions(self) -> list[SkyjoAction]:
+        actions = [
+            SkyjoAction(SkyjoActionType.DISCARD_AND_FLIP, row_idx, col_idx)
+            for row_idx, col_idx in self.face_down_indices
+        ]
+        return actions
+
+    @property
+    def valid_place_from_discard_actions(self) -> list[SkyjoAction]:
+        actions = [
+            SkyjoAction(SkyjoActionType.PLACE_FROM_DISCARD, row_idx, col_idx)
+            for row_idx, col_idx in self.visible_indices + self.face_down_indices
+        ]
+        return actions
+
+    @property
+    def valid_place_drawn_actions(self) -> list[SkyjoAction]:
+        actions = [
+            SkyjoAction(SkyjoActionType.PLACE_DRAWN, row_idx, col_idx)
+            for row_idx, col_idx in self.visible_indices + self.face_down_indices
+        ]
+        return actions
+
+    @staticmethod
+    def clearable(cards: np.array, is_flipped: np.array, col_idx: int):
+        """This is a static method so we can check whether the updated Hand
+        we are about to create needs to have any cleared columns, preventing us
+        from creating any intermediate Hands with incorrect is_cleared values."""
+        all_same = True
+        for r in range(NUM_ROWS):
+            all_same = (
+                all_same
+                and is_flipped[Hand.flat_index(r, col_idx)]
+                and cards[Hand.flat_index(0, col_idx)]
+                == cards[Hand.flat_index(r, col_idx)]
+            )
+        return all_same
+
+    def has_all_revealed(self) -> bool:
+        # To be cleared the is_flipped value would also be set to 1, so we can just
+        # count how many flipped cards there are without worrying about cleared counts.
+        return self.is_flipped.sum() == NUM_COLUMNS * NUM_ROWS
```

Comment:
I think u could probably just do `all(self.is_flipped)` and get short-circuiting